File: PlayerDemo/DTPlayerControlHandler/DTPlayerControlHandler.m
//
//  DTPlayerControlHandler.m
//  PlayerDemo
//
//  Created by Eden on 2022/12/30.
//

@import UIKit;
#import "DTPlayerControlHandler.h"

typedef NS_ENUM(NSInteger, DTPlayerControlEvent){
    
    DTPlayerControlDefaultEvent = 0,
    
    DTPlayerControlVolumeEvent = 1,
    
    DTPlayerControlSeekEvent = 2,
};

#pragma mark - DTPlayerControlConfiguration -

@interface DTPlayerControlConfiguration (Private)

+ (instancetype)defaultConfiguration;

@end

#pragma mark - DTPlayerControlHandler -

@interface DTPlayerControlHandler ()

@property (strong, nonatomic) DTPlayerControlConfiguration *configuration;

@property (assign, nonatomic) DTPlayerControlEvent event;

@property (weak, nonatomic) UIGestureRecognizer *gestureRecognizer;

@end

@implementation DTPlayerControlHandler

+ (instancetype)controlHandlerWithDelegate:(id<DTPlayerControlHandlerDelegate>)delegate
{
    DTPlayerControlHandler *__autoreleasing handler = [[DTPlayerControlHandler alloc] init];
    
    return handler;
}

- (instancetype)init
{
    self = [super init];
    
    if (self == nil) {
        
        return nil;
    }
    
    DTPlayerControlConfiguration *configuration = [DTPlayerControlConfiguration defaultConfiguration];
    
    [self setConfiguration:configuration];
    
    return self;
}

- (void)setEnabled:(BOOL)enabled
{
    self.gestureRecognizer.enabled = enabled;
}

- (BOOL)isEnabled
{
    return self.gestureRecognizer.enabled;
}

- (void)addToView:(UIView *)view
{
    if (self.gestureRecognizer != nil) {
        
        UIGestureRecognizer *gestureRecognizer = self.gestureRecognizer;
        
        // Remove from previous view.
        UIView *previousView = gestureRecognizer.view;
        [previousView removeGestureRecognizer:gestureRecognizer];
        
        // Add to new view.
        [view addGestureRecognizer:gestureRecognizer];
        return;
    }
    
    UIPanGestureRecognizer *gestureRecognizer = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(panGestureHandlerWithSender:)];
    
    [view addGestureRecognizer:gestureRecognizer];
    [self setGestureRecognizer:gestureRecognizer];
}

#pragma mark - Actions -

- (void)panGestureHandlerWithSender:(UIPanGestureRecognizer *)sender
{
    UIGestureRecognizerState state = sender.state;
    UIView *view = sender.view;
    CGPoint velocity = [sender velocityInView:view];
    
    if (state != UIGestureRecognizerStateBegan) {
        
        // When not begin state.
        if (self.event == DTPlayerControlVolumeEvent) {
            
            [self volumeGestureHandlerWithSender:sender];
        } else {
            
            [self seekGestureHandlerWithSender:sender];
        }
        
        return;
    }
    
    // Begin state.
    CGFloat maximumVelocityOfDirection = fmax(fabs(velocity.x), fabs(velocity.y));
    DTPlayerControlEvent event = (maximumVelocityOfDirection == fabs(velocity.x)) ? DTPlayerControlSeekEvent : DTPlayerControlVolumeEvent;
    
    // Notify delegate
    if (event == DTPlayerControlVolumeEvent) {
        
        [self notifyWillChangeVolumeLevelDelegate];
    }
    
    if (event == DTPlayerControlSeekEvent) {
        
        [self notifyWillBeginSeekDelegate];
    }
    
    [self setEvent:event];
}

- (void)volumeGestureHandlerWithSender:(UIPanGestureRecognizer *)sender
{
    UIGestureRecognizerState state = sender.state;
    UIView *view = sender.view;
    CGPoint point = [sender translationInView:view];
    CGFloat volumeLevel = self.volumeLevel;
    
    if (state == UIGestureRecognizerStateEnded) {
        
        [self setEvent:DTPlayerControlDefaultEvent];
        [self notifyWillChangeVolumeLevelDelegate];
    }
    
    if (state != UIGestureRecognizerStateChanged) {
        
        return;
    }
    
    CGFloat volumeGestureScale = self.configuration.gestureScale.volume;
    
    if (fabs(point.y) >= volumeGestureScale) {
        
        CGFloat multiple = fabs(point.y) / volumeGestureScale;
        CGFloat volumeIncrement = self.configuration.increament.volume;
        CGFloat addtionInterval = (point.y > 0.0f) ? -volumeIncrement : volumeIncrement;
        
        volumeLevel += (addtionInterval * multiple);
        volumeLevel = fmin(volumeLevel, 1.0f);
        volumeLevel = fmax(0.0, volumeLevel);
        
        [self setVolumeLevel:volumeLevel];
        [self notifyDidChangeVolumeLevelDelegate:volumeLevel];
        [sender setTranslation:CGPointZero inView:view];
    }
}

- (void)seekGestureHandlerWithSender:(UIPanGestureRecognizer *)sender
{
    UIGestureRecognizerState state = sender.state;
    UIView *view = sender.view;
    CGPoint point = [sender translationInView:view];
    CGFloat seekingInterval = self.playedInterval;
    
    if (state == UIGestureRecognizerStateEnded) {
        
        [self setEvent:DTPlayerControlDefaultEvent];
        [self notifyDidFinishSeekToSecondsDelegate:seekingInterval];
    }
    
    if (state != UIGestureRecognizerStateChanged) {
        
        return;
    }
    
    CGFloat seekGestureScale = self.configuration.gestureScale.seek;
    
    if (fabs(point.x) >= seekGestureScale) {
        
        CGFloat multiple = fabs(point.x) / seekGestureScale;
        CGFloat seekIncrement = self.configuration.increament.seek;
        CGFloat addtionInterval = (point.x > 0.0f) ? seekIncrement : -seekIncrement;
        
        seekingInterval += (addtionInterval * multiple);
        seekingInterval = fmin(seekingInterval, self.totalPlayInterval);
        seekingInterval = fmin(0.0f, seekingInterval);
        
        [self setPlayedInterval:seekingInterval];
        [self notifyDidSeekToSecondsDelegate:seekingInterval];
        [sender setTranslation:CGPointZero inView:view];
    }
}

#pragma mark - Notify Delegate -

- (void)notifyWillChangeVolumeLevelDelegate
{
    if (![self.delegate respondsToSelector:@selector(controlHandlerWillChangeVolumeLevel:)]) {
        
        return;
    }
    
    [self.delegate controlHandlerWillChangeVolumeLevel:self];
}

- (void)notifyDidChangeVolumeLevelDelegate:(CGFloat)level
{
    if (![self.delegate respondsToSelector:@selector(controlHandler:didChangeVolumeLevel:)]) {
        
        return;
    }
    
    [self.delegate controlHandler:self didChangeVolumeLevel:level];
}

- (void)notifyWillBeginSeekDelegate
{
    if (![self.delegate respondsToSelector:@selector(controlHandlerWithDelegate:)]) {
        
        return;
    }
    
    [self.delegate controlHandlerWillBeginSeek:self];
}

- (void)notifyDidSeekToSecondsDelegate:(CGFloat)seconds
{
    if (![self.delegate respondsToSelector:@selector(controlHandler:didSeekToSeconds:)]) {
        
        return;
    }
    
    [self.delegate controlHandler:self didSeekToSeconds:seconds];
}

- (void)notifyDidFinishSeekToSecondsDelegate:(CGFloat)seconds
{
    if (![self.delegate respondsToSelector:@selector(controlHandler:didFinishSeekToSeconds:)]) {
        
        return;
    }
    
    [self.delegate controlHandler:self didFinishSeekToSeconds:seconds];
}

@end

File: PlayerDemo/DTPlayerControlHandler.swift
//
//  DTPlayerControlHandler.swift
//
//  Created by Darktt on 19/9/25.
//  Copyright © 2019 Darktt. All rights reserved.
//

import Foundation
import UIKit.UIGestureRecognizer
import CoreGraphics.CGBase

public class DTPlayerControlHandler
{
    // MARK: - Properties -
    
    public weak var delegate: DTPlayerControlHandlerDelegate?
    
    public var configuration: Configuration = Configuration()
    
    public var volumeLevel: CGFloat = 0.0
    
    public var totalPlayInterval: TimeInterval = 0.0
    
    public var playedInterval: TimeInterval = 0.0
    
    public var isEnabled: Bool = true {
        
        willSet {
            
            self.gestureRecognizer?.isEnabled = newValue
        }
    }
    
    private var event: Event = .default
    private var gestureRecognizer: UIGestureRecognizer?
    
    // MARK: - Methods -
    // MARK: Initial Method
    
    public init(delegate: DTPlayerControlHandlerDelegate)
    {
        self.delegate = delegate
    }
    
    public func add(to view: UIView)
    {
        if let gestureRecognizer = self.gestureRecognizer {
            
            // Remove from previous view.
            let previousView: UIView = gestureRecognizer.view!
            previousView.removeGestureRecognizer(gestureRecognizer)
            
            // Add to new view.
            view.addGestureRecognizer(gestureRecognizer)
            return
        }
        
        let gestureRecognizer = UIPanGestureRecognizer(target: self, action: #selector(panGestureHandler(_:)))
        
        view.addGestureRecognizer(gestureRecognizer)
        self.gestureRecognizer = gestureRecognizer
    }
}

// MARK: - Actions -

private extension DTPlayerControlHandler
{
    @objc
    func panGestureHandler(_ sender: UIPanGestureRecognizer)
    {
        let state: UIGestureRecognizer.State = sender.state
        let view: UIView? = sender.view
        let velocity: CGPoint = sender.velocity(in: view)
        
        if state != .began {
            // When not begin state.
            if self.event == .volume {
                
                self.volumeGestureHandler(sender)
            } else {
                
                self.seekGestureHandler(sender)
            }
            return
        }
        
        // Begin state.
        let maximumVelocityOfDirection: CGFloat = max(abs(velocity.x), abs(velocity.y))
        let event: Event = (maximumVelocityOfDirection == abs(velocity.x)) ? .seek : .volume
        
        // Notify delegate
        if event == .volume {
            
            self.delegate?.controlHandler(willChangeVolumeLevel: self)
        }
        
        if event == .seek {
            
            self.delegate?.controlHandler(willBeginSeek: self)
        }
        
        self.event = event
    }
    
    func volumeGestureHandler(_ sender: UIPanGestureRecognizer)
    {
        let state: UIGestureRecognizer.State = sender.state
        let view: UIView? = sender.view
        let point: CGPoint = sender.translation(in: view)
        var volumeLevel: CGFloat = self.volumeLevel
        
        if state == .ended {
            
            self.event = .default
            self.volumeLevel = 0.0
            self.delegate?.controlHandler(self, didChangeVolumeLevel: volumeLevel)
        }
        
        if state != .changed {
            
            return
        }
        
        let volumeGestureScale: CGFloat = self.configuration.gestureScale.volume
        
        if abs(point.y) >= volumeGestureScale {
            
            let multiple: CGFloat = abs(point.y) / volumeGestureScale
            let volumeIncrement: CGFloat = self.configuration.increment.volume
            let addtionInterval: CGFloat = (point.y > 0.0) ? -volumeIncrement : volumeIncrement
            
            volumeLevel += (addtionInterval * multiple)
            
            volumeLevel = min(volumeLevel, 1.0)
            volumeLevel = max(0.0, volumeLevel)
            
            self.volumeLevel = volumeLevel
            self.delegate?.controlHandler(self, didChangeVolumeLevel: volumeLevel)
            sender.setTranslation(.zero, in: view)
        }
    }
    
    func seekGestureHandler(_ sender: UIPanGestureRecognizer)
    {
        let state: UIGestureRecognizer.State = sender.state
        let view: UIView? = sender.view
        let point: CGPoint = sender.translation(in: view)
        var seekingInterval: TimeInterval = self.playedInterval
        
        if state == .ended {
            
            self.event = .default
            self.delegate?.controlHandler(self, didFinishSeekToSeconds: seekingInterval)
        }
        
        if state != .changed {
            
            return
        }
        
        let seekGestureScale: CGFloat = self.configuration.gestureScale.seek
        
        if abs(point.x) >= seekGestureScale {
            
            let multiple: CGFloat = abs(point.x) / seekGestureScale
            let seekIncrement: CGFloat = self.configuration.increment.seek
            let addtionInterval: CGFloat = (point.x > 0.0) ? seekIncrement : -seekIncrement
            
            seekingInterval += TimeInterval(addtionInterval * multiple)
            seekingInterval = min(seekingInterval, self.totalPlayInterval)
            seekingInterval = max(0.0, seekingInterval)
            
            self.playedInterval = seekingInterval
            self.delegate?.controlHandler(self, didSeekToSeconds: seekingInterval)
            sender.setTranslation(.zero, in: view)
        }
    }
}

// MARK: - DTPlayerControlHandler.Configuration -

public extension DTPlayerControlHandler
{
    struct Configuration
    {
        var gestureScale: Group = Group()
        
        var increment: Group = Group()
        
        fileprivate init()
        {
            self.gestureScale.volume = 2.0
            self.gestureScale.seek = 1.0
            
            self.increment.volume = 0.01
            self.increment.seek = 1.0
        }
    }
}

// MARK: - DTPlayerControlHandler.Configuration.Group -

public extension DTPlayerControlHandler.Configuration
{
    struct Group {
        
        var volume: CGFloat = 0.0
        
        var seek: CGFloat = 0.0
        
        fileprivate init() {}
    }
}

// MARK: - DTPlayerControlHandler.Event -

private extension DTPlayerControlHandler
{
    enum Event
    {
        case `default`
        
        case volume
        
        case seek
    }
}

// MARK: - DTPlayerControlHandlerDelegate -

public protocol DTPlayerControlHandlerDelegate: AnyObject
{
    func controlHandler(willChangeVolumeLevel handler: DTPlayerControlHandler)
    
    func controlHandler(_ handler: DTPlayerControlHandler, didChangeVolumeLevel level: CGFloat)
    
    func controlHandler(willBeginSeek handler: DTPlayerControlHandler)
    
    func controlHandler(_ handler: DTPlayerControlHandler, didSeekToSeconds seconds: TimeInterval)
    
    func controlHandler(_ handler: DTPlayerControlHandler, didFinishSeekToSeconds seconds: TimeInterval)
}
